File: app/src/main/java/skripsi/kohonen/signaturerecognition/Recognition.java
package skripsi.kohonen.signaturerecognition;

import android.graphics.Bitmap;
import android.util.Log;

import java.util.ArrayList;

import butterknife.InjectView;

public class Recognition {
    private Database db;
    private Kohonen kh;
    public static boolean index= false;
    public static boolean index2= false;


    public Recognition(Database db) {
        this.db = db;
        kh = new Kohonen(db);

    }

    public void hapusSemua() {
        db.hapusSemua();
        if (kh != null) {
            kh.hapusSemua();
        }
    }

    @InjectView(R.id.drawing_simpan)
    FiturDrawing drawing;
    @InjectView(R.id.drawing_uji)
    FiturDrawing drawing2;
    public void setDrawing(FiturDrawing drawing) {
        this.drawing = drawing;
        this.drawing2= drawing;
    }

    public Bitmap ambilGambarDrawing_simpan() {
        drawing.setDrawingCacheEnabled(true);
        drawing.buildDrawingCache(true);
        Bitmap gambar = Bitmap.createBitmap(drawing.getDrawingCache());
        drawing.setDrawingCacheEnabled(false);
        return gambar;
    }
    public boolean cekArr = true;
    public Bitmap ambilGambarDrawing_uji() {

        drawing2.setDrawingCacheEnabled(true);
        drawing2.buildDrawingCache(true);
        Bitmap gambar2 = Bitmap.createBitmap(drawing2.getDrawingCache());
        drawing2.setDrawingCacheEnabled(false);
        return gambar2;
    }

    public ArrayList<String>  tes() {
       Bitmap gambar2 = ImageProcessing.proses(ambilGambarDrawing_uji());
        byte[] arrImage = ImageProcessing.getArrImage(gambar2);

        ArrayList<String> i = kh.tesDataBaru(arrImage);
        ambilJarakMaksimal(i.get(0));
        return i;
    }

    public static int nilaiMax = 0;
    public void ambilJarakMaksimal(String nama) {
        double jarak = 0;
        double hasil= 0;
        ArrayList<TandaTangan> arrTtd = db.ambilDataBerdasarkanNama(nama);
        byte[][] arrBiner = new byte[arrTtd.size()][40000];
        ArrayList<Double> arr;
        /*Log.d("size: ",String.valueOf(arrTtd.size()));*/
        for (int i = 0; i<arrTtd.size(); i++) {
            arrBiner[i] = TandaTangan.strToArr(arrTtd.get(i).getNilai());
        }
        arr = kh.tesBerdasarkanNama(arrBiner);

        nilaiMax = maksimal(arr);
        jarak = ((25000-nilaiMax)/(0.65)-25000)*-1;
        hasil = (25000-Kohonen.minimal)/(25000-jarak)*100;
        Log.d("hasil: ", String.valueOf(hasil));
        if (hasil <0) {
            Kohonen.asli = 0;
        }else if (hasil >100){
            Kohonen.asli = 100;
        }else
        Kohonen.asli = hasil;
        Log.d("nilai max", String.valueOf(nilaiMax));
        //return jarak;
    }

    private int maksimal(ArrayList<Double> nodeArray) {
        double max = nodeArray.get(0);
        for (int i = 0; i < nodeArray.size(); i++) {
            if (nodeArray.get(i) > max) {
                max = nodeArray.get(i);
            }
        }
        return (int) max;
    }


    /*public byte[] cekuji(){
        Bitmap gambar2 = ImageProcessing.proses(ambilGambarDrawing_uji());
        byte[] arrIm = ImageProcessing.getArrImage(gambar2);
        return arrIm;
    }*/
    public void cekcoretan(){
        index = false;
        Bitmap gambar2 = ImageProcessing.proses(ambilGambarDrawing_uji());
        byte[] arrImage = ImageProcessing.getArrImage(gambar2);
        for (int i=0; i<arrImage.length; i++) {
            if (arrImage[i] == 1) {
                index = true;
                break;
            }
        }

    }
public void cek(){
    index2 =false;
    Bitmap gambar = ImageProcessing.proses(ambilGambarDrawing_simpan());
    byte[] arrImg = ImageProcessing.getArrImage(gambar);
    for (int i=0; i<arrImg.length; i++) {
        if (arrImg[i] == 1) {
            index2 = true;
            break;
        }
    }
}
    public void simpan(String str) {

        Bitmap gambar = ImageProcessing.proses(ambilGambarDrawing_simpan());
        byte[] arrImg = ImageProcessing.getArrImage(gambar);
        db.addttd(new TandaTangan(db.getLastIndex()+1, str, TandaTangan.arrToStr(arrImg)));
        kh.latihDataBaru(arrImg, str);
    }

    public void hapus(int pos) {
        db.deletettd(pos);
        kh.hapus(pos);
    }

    public void ubah(int pos, String nama) {
        TandaTangan kat = db.getttd(pos);
        kat.setnama(nama);
        db.updatettd(kat);
        kh.ubah(pos, nama);
    }

    public byte[] getArrDataImage(int i) {
        return TandaTangan.strToArr(db.getttd(i).getNilai());
    }

    public ArrayList<String> getDatanama() {
        return kh.huruf;
    }

    public void tampilData() {


    }

    public void tampilDb() {
        ArrayList<TandaTangan> list = db.getAllttd();
        for (TandaTangan kat : list) {
            Log.d("data ke-", kat.getId() + ", " + kat.getnama());
        }
        Log.d("selesai", "-");
    }
}

File: app/src/main/java/skripsi/kohonen/signaturerecognition/Recognition.kt
package skripsi.kohonen.signaturerecognition

import android.util.Log
import butterknife.InjectView
import java.util.*

class Recognition(private val db: Database?) {
    private val kh: Kohonen?
    fun hapusSemua() {
        db!!.hapusSemua()
        kh?.hapusSemua()
    }

    @InjectView(R.id.drawing_simpan)
    var drawing: FiturDrawing? = null

    @InjectView(R.id.drawing_uji)
    var drawing2: FiturDrawing? = null
    fun setDrawing(drawing: FiturDrawing?) {
        this.drawing = drawing
        drawing2 = drawing
    }

    fun ambilGambarDrawing_simpan(): Bitmap {
        drawing!!.isDrawingCacheEnabled = true
        drawing!!.buildDrawingCache(true)
        val gambar: Bitmap = Bitmap.createBitmap(drawing!!.drawingCache)
        drawing!!.isDrawingCacheEnabled = false
        return gambar
    }

    var cekArr = true
    fun ambilGambarDrawing_uji(): Bitmap {
        drawing2!!.isDrawingCacheEnabled = true
        drawing2!!.buildDrawingCache(true)
        val gambar2: Bitmap = Bitmap.createBitmap(drawing2!!.drawingCache)
        drawing2!!.isDrawingCacheEnabled = false
        return gambar2
    }

    fun tes(): ArrayList<String?>? {
        val gambar2: Bitmap = ImageProcessing.Companion.proses(ambilGambarDrawing_uji())
        val arrImage: ByteArray = ImageProcessing.Companion.getArrImage(gambar2)
        val i = kh!!.tesDataBaru(arrImage)
        ambilJarakMaksimal(i!![0])
        return i
    }

    fun ambilJarakMaksimal(nama: String?) {
        var jarak = 0.0
        var hasil = 0.0
        val arrTtd = db!!.ambilDataBerdasarkanNama(nama)
        val arrBiner = Array(arrTtd!!.size) { ByteArray(40000) }
        val arr: ArrayList<Double?>?
        /*Log.d("size: ",String.valueOf(arrTtd.size()));*/for (i in arrTtd.indices) {
            arrBiner[i] = TandaTangan.Companion.strToArr(arrTtd[i]!!.nilai)
        }
        arr = kh!!.tesBerdasarkanNama(arrBiner)
        nilaiMax = maksimal(arr)
        jarak = ((25000 - nilaiMax) / 0.65 - 25000) * -1
        hasil = (25000 - Kohonen.Companion.minimal) / (25000 - jarak) * 100
        Log.d("hasil: ", hasil.toString())
        if (hasil < 0) {
            Kohonen.Companion.asli = 0.0
        } else if (hasil > 100) {
            Kohonen.Companion.asli = 100.0
        } else Kohonen.Companion.asli = hasil
        Log.d("nilai max", nilaiMax.toString())
        //return jarak;
    }

    private fun maksimal(nodeArray: ArrayList<Double>?): Int {
        var max = nodeArray!![0]
        for (i in nodeArray.indices) {
            if (nodeArray[i] > max) {
                max = nodeArray[i]
            }
        }
        return max.toInt()
    }

    /*public byte[] cekuji(){
        Bitmap gambar2 = ImageProcessing.proses(ambilGambarDrawing_uji());
        byte[] arrIm = ImageProcessing.getArrImage(gambar2);
        return arrIm;
    }*/
    fun cekcoretan() {
        index = false
        val gambar2: Bitmap = ImageProcessing.Companion.proses(ambilGambarDrawing_uji())
        val arrImage: ByteArray = ImageProcessing.Companion.getArrImage(gambar2)
        for (i in arrImage.indices) {
            if (arrImage[i] == 1) {
                index = true
                break
            }
        }
    }

    fun cek() {
        index2 = false
        val gambar: Bitmap = ImageProcessing.Companion.proses(ambilGambarDrawing_simpan())
        val arrImg: ByteArray = ImageProcessing.Companion.getArrImage(gambar)
        for (i in arrImg.indices) {
            if (arrImg[i] == 1) {
                index2 = true
                break
            }
        }
    }

    fun simpan(str: String?) {
        val gambar: Bitmap = ImageProcessing.Companion.proses(ambilGambarDrawing_simpan())
        val arrImg: ByteArray = ImageProcessing.Companion.getArrImage(gambar)
        db!!.addttd(TandaTangan(db.lastIndex + 1, str, TandaTangan.Companion.arrToStr(arrImg)))
        kh!!.latihDataBaru(arrImg, str)
    }

    fun hapus(pos: Int) {
        db!!.deletettd(pos)
        kh!!.hapus(pos)
    }

    fun ubah(pos: Int, nama: String?) {
        val kat = db!!.getttd(pos)
        kat!!.setnama(nama)
        db.updatettd(kat)
        kh!!.ubah(pos, nama)
    }

    fun getArrDataImage(i: Int): ByteArray {
        return TandaTangan.Companion.strToArr(db!!.getttd(i).nilai)
    }

    fun getDatanama(): ArrayList<String?>? {
        return kh!!.huruf
    }

    fun tampilData() {}
    fun tampilDb() {
        val list = db!!.allttd
        for (kat in list!!) {
            Log.d("data ke-", kat!!.id.toString() + ", " + kat!!.getnama())
        }
        Log.d("selesai", "-")
    }

    companion object {
        var index = false
        var index2 = false
        var nilaiMax = 0
    }

    init {
        kh = Kohonen(db)
    }
}